test/acceptance/stubba_example_test.rb

```diff
@@ -88,9 +88,9 @@ def should_stub_two_different_class_methods
     found_widgets = [Widget.new]
     created_widget = Widget.new
     Widget.expects(:find).with(:all).returns(found_widgets)
-    Widget.expects(:create).with(:model => 'wombat').returns(created_widget)
+    Widget.expects(:create).with({ :model => 'wombat' }).returns(created_widget)
     assert_equal found_widgets, Widget.find(:all)
-    assert_equal created_widget, Widget.create(:model => 'wombat')
+    assert_equal created_widget, Widget.create({ :model => 'wombat' })
```

Comment:
Hmm, `Mock#method_missing` is not tagging the last hash as a ruby2_keywords_hash 🤔. Current suspicion is the way `InstanceMethod#define_new_method` passes the kwarg to method_missing is incompatible:
```ruby
stub_method_owner.send(:define_method, method_name) do |*args, &block|
        self_in_scope.mock.method_missing(method_name_in_scope, *args, &block)
      end
```

Reply:
Does it help if you add `**kwargs` to the block params and the `method_missing` params? i.e. 

```ruby
stub_method_owner.send(:define_method, method_name) do |*args, **kwargs, &block|
  self_in_scope.mock.method_missing(method_name_in_scope, *args, **kwargs, &block)
end
```

Reply:
Setting `ruby2_keywords` on the methods through the flow works (see https://github.com/freerange/mocha/pull/534/commits/96e71ad16d21606b2d6f3f32b24982ee28eda1f5), but there probably is another way that doesn't implicitly couple code there 🤔 At least the acceptance test passes now.

Reply:
Ah, gotcha! That blog post looks very helpful.

Reply:
`ruby2_keywords` seems to work in Ruby >=2.7 (and <2.6 with a patch), so I don't know if we need to conditionally enable it only on Ruby 3 as well 🤞.

Reply:
Was discussing with @chrisseaton (thanks!), and he suggested to try using `ruby2_keywords` only at the "entry points" (e.g. `Mock#method_missing` and `StubbedMethod#define_new_method`). At that point, we can already split between positional and keyword arguments, which should make subsequent logic cleaner instead of having to set/check for ruby2_keywords at multiple layers (e.g. https://github.com/freerange/mocha/commit/96e71ad16d21606b2d6f3f32b24982ee28eda1f5).

It sounded like a great idea, so I started spiking a possible encapsulation of parameters [here](https://github.com/wasabigeek/mocha/pull/1/files#diff-5440fe2ba7746f3290444b3d79be4fa757b7c4b6d489cf6989a66ad478795a51R31-R32), but it's looking to be a pretty large and tricky change.

@floehopper, how do you feel about the refactor? If agreeable, do you think we should attempt to refactor first, or should I try to finish this current approach first (setting ruby2_keywords at multiple layers) before circling back?

Reply:
The context is that the current approach unpacks arguments, and then packs them again, multiple times. That's going to be tricky and error-prone, so the idea was to do it just once, and then keep them as structured data beyond that.

Reply:
@wasabigeek @chrisseaton Thanks, both. I'm sorry that I don't have much time to help you with this at the moment, so I'm hugely grateful to you for working on it.

> @floehopper, how do you feel about the refactor? If agreeable, do you think we should attempt to refactor first, or should I try to finish this current approach first (setting ruby2_keywords at multiple layers) before circling back?

I'm beginning to understand that a bigger refactor of the parameter matchers might be the only sensible way forward and I'm completely open to that - it's not entirely surprising as the current design originated 15 years ago when Ruby v1.8.6 was the current release!

Ideally we'd try to keep as much of the public API as stable as possible, but if that's not possible then I'd be completely happy to either use a configuration option or (more likely) do a major version bump. Do let me know if you have any other questions or if there's anything you need from me to make progress!

I don't have a strong opinion about whether the refactoring is done first or not. If you're confident the refactoring will make making the actual change easier, then doing it first makes sense. If you still need to get confidence that the refactoring will actually help, then perhaps it's worthwhile to do some more exploration first.

Just FYI I've recently removed support for Ruby v1.8 and I'm quite keen to remove support for some of the older versions of Test::Unit & Minitest which required monkey-patching, so perhaps all these changes could all be included in a major version release...? I'm also open to removing support for some slightly newer versions of Ruby if that would make life a lot easier...

Reply:
@floehopper I've sketched out a refactor here, along with comments on how the future Ruby 3 changes could be incorporated: https://github.com/freerange/mocha/pull/544, appreciate your thoughts.

API-wise, I think there will be some ambiguity when hash matchers are involved e.g. is `.with(has_value(“hello”))` matching a last positional Hash, or a kwarg? I think we can possibly:
- try to allow hash matchers to check for both last positional Hash _and_ kwargs, passing if either scenario matches (which would be ambiguous but shouldn't break existing test suites, and maybe we can phase this out later in favour of explicitness)
- introduce a way to explicitly declare a kwarg matcher, e.g.:
    - `.with(arg, kwargs(has_value(“hello”)))`
    - `.with(arg, kwargs: has_value(“hello”))`

What do you think?

Reply:
> @floehopper I've sketched out a refactor here [...]

@wasabigeek I'm a bit confused where "here" is, i.e. where I should be looking! Is it on the branch associated with this PR? And if so what's the best way for me to look at it - should I just look at the net diff of all the changes or each commit in turn?

Reply:
Ah sorry, I was referencing another PR: https://github.com/freerange/mocha/pull/544, that has the proposed refactor.

Reply:
@wasabigeek:

> API-wise, I think there will be some ambiguity when hash matchers are involved e.g. is `.with(has_value(“hello”))` matching a last positional Hash, or a kwarg? I think we can possibly:
> 
> * try to allow hash matchers to check for both last positional Hash _and_ kwargs, passing if either scenario matches (which would be ambiguous but shouldn't break existing test suites, and maybe we can phase this out later in favour of explicitness)
> * introduce a way to explicitly declare a kwarg matcher, e.g.:
>   
>   * `.with(arg, kwargs(has_value(“hello”)))`
>   * `.with(arg, kwargs: has_value(“hello”))`
> 
> What do you think?

Ideally I'd like to provide as simple an upgrade path as possible given that this change might mean quite pervasive changes in a lot of codebases. I think that means having a backwardly compatible version of the code (ideally with deprecation warnings highlighting the ambiguities) and a new version of the code which requires explicitness. It would be nice if the two versions of the code could co-exist in the same release of Mocha behind a configuration option, but if that makes the code ridiculously complicated, it would be fine to have the two versions of the code in separate releases of Mocha (separated by a major version bump). In the latter case, it would be good to have finalised the new API by the time of the 1st release, so the deprecation warnings could explain how to fix them as clearly as possible.

I haven't had much time to think about the API, but of the two options you've proposed, I think I prefer `.with(arg, kwargs(has_value(“hello”)))` - it seems a bit more in keeping with the existing API. I guess another option might be to have a separate method on `Expectation`, e.g. `with_kwargs`.

Do you think we might need to make the positional argument matching more explicit too, i.e. `.with(args(arg1, arg2), kwargs(has_value(“hello”)))`...?

If we do use any of these approaches, I'd probably prefer more verbose names, i.e. `keyword_arguments` vs `kwargs`; maybe with aliases for the abbreviations...

Just to check my understanding, am I right in thinking the ambiguity mentioned above will eventually go away if/when we stop supporting Ruby v2? If that's the case, would it make any sense to have a configuration option (or code that detects the Ruby version) that allows a simpler `Expectation#with` signature (i.e. the same as the existing signature) when your using Ruby v3...? Apologies if I've misunderstood this!

I hope that helps. Thanks again for your work on this. Let me know if you have any more questions.

Reply:
Sorry, I forgot to respond to this!

> I guess another option might be to have a separate method on Expectation, e.g. with_kwargs.
> Do you think we might need to make the positional argument matching more explicit too, i.e. .with(args(arg1, arg2), kwargs(has_value(“hello”)))...?

I don't have a strong opinion at the moment, though it feels like `args(...)` wrapper changes the API without adding much. `with_kwargs` could be a more straightforward alternative 🤔 

> Just to check my understanding, am I right in thinking the ambiguity mentioned above will eventually go away if/when we stop supporting Ruby v2?

I don't think so, unless we make a breaking change in `with`'s signature (i.e. the ambiguity is more from mocha's existing API than from Ruby itself). The tricky part is determining whether something like `.expects(:foo).with(has_value(1))` is trying to exclusively match a last positional hash (e.g. `.foo({ a: 1})`) or a kwarg (e.g. `.foo(a: 1)`).

Reply:
@wasabigeek 

> > I guess another option might be to have a separate method on Expectation, e.g. with_kwargs.
> > Do you think we might need to make the positional argument matching more explicit too, i.e. .with(args(arg1, arg2), kwargs(has_value(“hello”)))...?
> 
> I don't have a strong opinion at the moment, though it feels like `args(...)` wrapper changes the API without adding much. `with_kwargs` could be a more straightforward alternative 🤔

Fair point.

> > Just to check my understanding, am I right in thinking the ambiguity mentioned above will eventually go away if/when we stop supporting Ruby v2?
> 
> I don't think so, unless we make a breaking change in `with`'s signature (i.e. the ambiguity is more from mocha's existing API than from Ruby itself). The tricky part is determining whether something like `.expects(:foo).with(has_value(1))` is trying to exclusively match a last positional hash (e.g. `.foo({ a: 1})`) or a kwarg (e.g. `.foo(a: 1)`).

Ah, I think I see what you mean now - thanks for explaining. 👍